pilot/pkg/config/clusterregistry/conversion.go

```diff
@@ -101,62 +100,140 @@ func (cs *ClusterStore) GetPilotClusters() (clusters []*k8s_cr.Cluster) {
+				log.Warnf("%v", err)
+				return nil, err
+			}
+			cs.cfgStore = cluster
+		} else {
+			cs.clusters = append(cs.clusters, cluster)
 		}
-		if !supportedExtensions[filepath.Ext(path)] || (info.Mode()&os.ModeType) != 0 {
-			return nil
+	}
+	if cs.cfgStore == nil {
+		log.Warnf("no config store is defined in the cluster registries")
+		return nil, nil
+	}
+
+	return
+}
+
+// getClustersConfigs(configMapName,configMapNamespace)
+func getClustersConfigs(k8s kubernetes.Interface, configMapName, configMapNamespace string, registryDir string) ([]*k8s_cr.Cluster, error) {
+	clusters := []*k8s_cr.Cluster{}
+	clusterRegistry, err := k8s.CoreV1().ConfigMaps(configMapNamespace).Get(configMapName, metav1.GetOptions{})
+	if err != nil {
+		return []*k8s_cr.Cluster{}, err
+	}
+	obj := k8s_cr.Cluster{}
+	for key, data := range clusterRegistry.Data {
+		fmt.Printf("key: %s data: %+v\n", key, data)
+		decoder := yaml.NewYAMLOrJSONDecoder(strings.NewReader(data), 4096)
+		err := decoder.Decode(&obj)
+		if err != nil {
+			log.Errorf("failed to decode cluster definition for: %s error: %v", key, err)
+			// TODO should fail here or go through the whole list
+			continue
 		}
-		data, err := ioutil.ReadFile(path)
+		err = validateCluster(&obj)
 		if err != nil {
-			log.Warnf("Failed to read %s: %v", path, err)
-			return err
+			log.Errorf("failed to validate cluster: %s error: %v", key, err)
+			// TODO should fail here or go through the whole list
+			continue
+		}
+		secretName := obj.ObjectMeta.Annotations[ClusterAccessConfigSecret]
+		secretNamespace := obj.ObjectMeta.Annotations[ClusterAccessConfigSecretNamespace]
+		if len(secretName) == 0 {
+			log.Errorf("cluster %s does not have annotation for Secret", key)
+			// TODO should fail here or go through the whole list
+			continue
+		}
+		if len(secretNamespace) == 0 {
+			secretNamespace = "istio-system"
 		}
-		result, err := parseClusters(crPath, data)
+		kubeconfig, err := getClusterConfigFromSecret(k8s, secretName, secretNamespace, key)
 		if err != nil {
-			log.Warnf("Failed to parse cluster file %s: %v", path, err)
-			return err
+			log.Errorf("failed to get Secret %s in namespace %s for cluster %s with error: %v",
+				secretName, secretNamespace, key, err)
+			continue
 		}
-		clusters = append(clusters, result...)
-		return nil
-	})
-
-	if err == nil {
-		// For the time being, assume that ClusterPilotCfgStore is only set for one cluster only.
-		// This cluster will be used as the pilot's config store.
-		for _, cluster := range clusters {
-			log.Infof("ClusterPilotCfgStore: %s", cluster.ObjectMeta.Annotations[ClusterPilotCfgStore])
-			if isCfgStore, _ := strconv.ParseBool(cluster.ObjectMeta.Annotations[ClusterPilotCfgStore]); isCfgStore {
-				if cs.cfgStore != nil {
-					err = fmt.Errorf("multiple cluster config stores are defined")
-					log.Warnf("%v", err)
-					return nil, err
+		// Now kubeconf can be saved in regsitryDir for further processing
```

Comment:
So the secrets are read and then saved into a directory. How would this improve the security? The secrets can be mounted automatically when starting the pilot pod.

Reply:
I don't think we can mount them automatically. We could if they were all in a single secret ( one big kubeconfig ) - that may simplify the code further. But dynamically adding a cluster will not work well.

Reply:
@baodongli Very good question. If you check the code, you will see that they are stored in /tmp folder which is when run "in-cluster" mode is inside of a container and is not easily accessible. When you run out of cluster, then it would be most likely for development so we do not care.
Ideally it should not be done, but after going through the code I realized that there are many references to kubeconf as files and refactor all these would be a huge change which probably will not be accepted. Happy to discuss it more.

Reply:
@baodongli secrets are not mounted automatically to the pod, the only thing you get from free from k8s is service account token, which is mounted to every container running in the pod.
/var/run/secrets/kubernetes.io/serviceaccount.

Secrets will be mounted only if you make a reference to it in volumes.

Reply:
Agreed we want to enable dynamic expansion of multicluster - seems like the natural case.

Reply:
Ya, I'd like to discuss a bit.  I took a look at master, and not sure where all the kubeconfigfiles are being used immediately.  In memory is ideal - if it takes a couple more PRs to get there that seems ok.  Mind providing a few links to files outside `pilot/pkg/bootstrap/server.go` and `pilot/pkg/config/clusterregistry`?  If there are a slew, maybe its better left alone - otherwise I think server.go and clusterregistry files can easily be modified.

Reply:
Sorry, I meant to say mount secret volume.

With kubeconfig configured as secret, can you explain how this would increase the security? thanks

Reply:
The way it is done now, you have a configmap, where ALL kubeconfigs are listed in clear text, basically the access to configmaps in a namespace is less protected as more clients needs to access them. Secrets on the other hand are at least based64 encoded and I think in general should be used for security related things. Does it make sense? Please ping me on the slack if you want to chat more about it.

Reply:
the kubeconfig in the cluster registry configmap is also base64 encoded. 

If both the secrets and the configmaps are in the same namespace, then they are equally accessible. If not, the deployer of istio control plane should have access to both namespaces. If anyone can gain access to one of the namespace, I don't see why he wouldn't be able to another one. 

istio ca also uses secret mount to mTLS, which is supposedly to make service-service interaction secure. If someone gains access to the pilot's namespace (which the cluster registry config map resides now), then those secrets are accessible too.

Sorry for my ignorance here.

Reply:
That is why I introduced additional namespace parameter which defines where these secrets would be stored, jumping namespaces is way more difficult that just hacking objects in the same namespace. So for those with security concern, they can place these secrets into a special name space and how more stringent access control which process can list/get them.